### ISSUE TITLE
Better sentiment analysis

### ISSUE THREAD

**Issue by maxbbraun** [maxbbraun/trump2cash#3]
The sentiment analysis [sometimes gets it wrong](https://github.com/maxbbraun/trump2cash/blob/master/benchmark.md#124-2017-1249-pm-tuesday). It also assigns the tweet's overall sentiment to all companies, which may not be correct [in some cases](https://github.com/maxbbraun/trump2cash/blob/master/benchmark.md#118-2017-734-am-wednesday).

One idea would be to try targeted sentiment analysis, e.g. using the [IBM Watson AlchemyLanguage API](https://www.ibm.com/watson/developercloud/alchemy-language/api/v1/#targeted_sentiment).

**Comment by flaw600** [comment]
I've actually used Watson's API and would be happy to take a crack at it. Not sure at what pace that would be, but I'd be up for it

**Comment by tbranyen** [comment]
If you don't mind spawning a java jar and communicating with it over sockets, I worked on a project that did sentiment analysis on the 2012 presidential debate captions. You can see the implementation here (JS): https://github.com/sosolimited/recon_backend/blob/master/named-entity/index.js#L36

Not sure if it'll be helpful, but might be a little easier to start with (but certainly not as powerful as Watson).

**Comment by maxbbraun** [comment]
Ok, I tried both the Watson [sentiment analysis](https://www.ibm.com/watson/developercloud/alchemy-language/api/v1/?python#sentiment) and the [targeted sentiment](https://www.ibm.com/watson/developercloud/alchemy-language/api/v1/?python#targeted_sentiment) APIs.

Both perform worse than the current implementation using the [GCNL](https://cloud.google.com/natural-language/) APIs.

The targeted one is wrong a lot, I suspect because it just doesn't have enough words in a short tweet to work well. The non-targeted one is frequently neutral where GCNL detects the correct sentiment.

**Comment by flaw600** [comment]
@maxbbraun can you share your test set of tweets and if possible your Watson implementation? My admittedly very limited testing is showing both types of sentiment being accurate

**Comment by maxbbraun** [comment]
@flaw600 Make the changes, run the [benchmark](https://github.com/maxbbraun/trump2cash/blob/master/benchmark.md), then look at the diff.

**Comment by flaw600** [comment]
Interesting. You're correct of course. In my limited testing using tweets not used in the benchmark I noticed Watson performing better (meaning more accurately than on this set, I didn't compare against GCNL) on his most recent tweets but you're right to stay on GCNL for now. Very intriguing

**Comment by labidus** [comment]
For sentiment analyst, since you are using python, neural network with LSTM could be a solution
Something like this : https://github.com/bentrevett/pytorch-sentiment-analysis
